src/dbos.ts

```diff
@@ -0,0 +1,781 @@
+      if (!DBOS.isInWorkflow()) {
+        throw new DBOSInvalidWorkflowTransitionError();
+      }
+      return (getCurrentDBOSContext() as WorkflowContext).getEvent(workflowID, key, timeoutSeconds);
+    }
+    return DBOS.executor.getEvent(workflowID, key, timeoutSeconds);
+  }
+
+  //////
+  // Decorators
+  //////
+  static scheduled(schedulerConfig: SchedulerConfig) {
+    function scheddec<This, Return>(
+      target: object,
+      propertyKey: string,
+      inDescriptor: TypedPropertyDescriptor<(this: This, ...args: ScheduledArgs) => Promise<Return>>
+    ) {
+      const { descriptor, registration } = registerAndWrapContextFreeFunction(target, propertyKey, inDescriptor);
+      const schedRegistration = registration as unknown as SchedulerRegistrationBase;
+      schedRegistration.schedulerConfig = schedulerConfig;
+
+      return descriptor;
+    }
+    return scheddec;
+  }
+
+  static workflow(config: WorkflowConfig={})
+  {
+    function decorator <
+      This,
+      Args extends unknown[],
+      Return
+    >
+    (
+      target: object,
+      propertyKey: string,
+      inDescriptor: TypedPropertyDescriptor<(this: This, ...args: Args) => Promise<Return>>
+    )
+    {
+      const { descriptor, registration } = registerAndWrapContextFreeFunction(target, propertyKey, inDescriptor);
+      registration.workflowConfig = config;
+
+      const invokeWrapper = async function (this: This, ...rawArgs: Args): Promise<Return> {
+        const pctx = getCurrentContextStore();
+        let inst: ConfiguredInstance | undefined = undefined;
+        if (typeof this === 'function') {
+          // This is static
+        }
+        else {
+          inst = this as ConfiguredInstance;
+          if (!("name" in inst)) {
+            throw new DBOSInvalidWorkflowTransitionError();
+          }
+        }
+
+        let wfId = pctx?.idAssignedForNextWorkflow;
+
+        // If this is called from within a workflow, this is a child workflow,
+        //  For OAOO, we will need a consistent ID formed from the parent WF and call number
+        if (DBOS.isWithinWorkflow()) {
+          const wfctx = assertCurrentWorkflowContext();
+
+          const funcId = wfctx.functionIDGetIncrement();
+          wfId = wfId || (wfctx.workflowUUID + "-" + funcId);
+          const params: WorkflowParams = {
+            workflowUUID: wfId,
+            parentCtx: wfctx,
+            configuredInstance: inst,
+            queueName: pctx?.queueAssignedForWorkflows
+          };
+
+          const cwfh = await DBOSExecutor.globalInstance!.internalWorkflow(
+            registration.registeredFunction as unknown as WorkflowFunction<Args, Return>,
+            params, wfctx.workflowUUID, funcId, ...rawArgs
+          );
+          return await cwfh.getResult();
+        }
+
+        const wfParams: InternalWorkflowParams = {
+          workflowUUID: wfId,
+          queueName: pctx?.queueAssignedForWorkflows,
+          configuredInstance : inst
```

Comment:
We could decide to do all the `pctx` introspection here and create a new `WorkflowContextImpl` instance, configured with all we want, and set it as parentCtx. I think this would be cleaner than the current logic in dbos-executor.